filesystem/client_interface.go

```diff
@@ -0,0 +1,11 @@
+package filesystem
+
+import "io"
+
+type Client interface {
+	OpenReadCloser(name string) (io.ReadCloser, error)
+	OpenWriteCloser(name string) (io.WriteCloser, error)
+	Exists(name string) (bool, error)
+	Rename(oldpath, newpath string) error
+	Glob(dirname, pattern string) ([]string, error)
```

Comment:
The reason to use `dirname` is to simplify impl. Full pathname regex is hard, and expensive on HDFS.

> Patterns will be like `/aaa/bbb/etl*/part.*`

The current Glob() is enough to do it by two level globbing. I can help you write it.